www/docs/server/validators.md

```diff
@@ -0,0 +1,339 @@
+---
+id: validators
+title: Input & Output Validators
+sidebar_label: Input & Output Validators
+slug: /server/validators
+---
+
+tRPC procedures may define an input and/or an output. It's flexible too, so we support many different validators, and it's easy to integrate a validator which we don't directly support.
+
+### Input Validators
+
+By defining an input validator, tRPC can check that a procedure call is correct and return a validation error if not. Just use the `.input()` method when writing your procedures.
+
+```ts twoslash
+// @target: esnext
+import { initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+const publicProcedure = t.procedure
+
+// ---cut---
+
+export const appRouter = t.router({
+  hello: publicProcedure
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const name = opts.input.name;
+      //    ^?
+      return {
+        greeting: `Hello ${opts.input.name}`,
+      };
+    }),
+});
+```
+
+`.input` can also be stacked to build more complex types, which is particularly useful when you want to utilise some common input to a collection of procedures in a [middleware](middlewares).
+
+```ts twoslash
+// @target: esnext
+import { TRPCError, initTRPC } from '@trpc/server';
+import { z } from 'zod';
+
+export const t = initTRPC.create();
+
+// ---cut---
+
+const procedureWhichRejectsOutOfTownFolk = t.procedure
+  .input(z.object({ townName: z.string() }))
+  .use((opts) => {
+    if (opts.input.townName !== 'Pucklechurch') {
+      throw new TRPCError({
+        code: 'FORBIDDEN',
+        message: "We don't take kindly to out-of-town folk",
+      });
+    }
+
+    return opts.next();
+  });
+
+export const appRouter = t.router({
+  hello: procedureWhichRejectsOutOfTownFolk
+    .input(
+      z.object({
+        name: z.string(),
+      }),
+    )
+    .query((opts) => {
+      const input = opts.input;
+      //    ^?
+      return {
+        greeting: `Hello ${input.name}, my friend`,
+      };
+    }),
+});
+```
+
+### Output Validators
+
+By defining an output validator, tRPC can also check that data returned from untrusted sources is correct, or to strip extraneous data. Defining outputs is not always as important as defining inputs, since tRPC gives you automatic type-safety by inferring the return type of your procedures.
+
+Validators are used to infer the input and output types of your procedures.
+
+:::info
+If output validation fails, the server will respond with an `INTERNAL_SERVER_ERROR`.
+:::
+
+## Using a function as a validator
```

Comment:
maybe this section makes more sense after the integrations?

Reply:
I think I'm trying to angle it as an initial representation of what a validator is doing. I'll do another pass now and see if you feel the same (happy to move it, I considered both ways as a coin toss)